compiler/src/iree/compiler/Codegen/Common/ReconcileTranslationInfo.cpp

```diff
@@ -32,6 +35,272 @@ class ReconcileTranslationInfoPass final
 };
 } // namespace
 
+//===---------------------------------------------------------------------===//
+// Resolve `scf.forall` operations
+//===---------------------------------------------------------------------===//
+
+/// Verify that the mapping attribute provided is of the right form.
+static FailureOr<SmallVector<IREE::Codegen::WorkgroupMappingAttr>>
+verifyWorkgroupMappingAttrArray(scf::ForallOp forallOp) {
+  std::optional<ArrayAttr> mappingAttr = forallOp.getMapping();
+  if (!mappingAttr) {
+    return forallOp.emitOpError("expected mapped for all op");
+  }
+  if (mappingAttr.value().empty()) {
+    return forallOp.emitOpError("mapping attribute cannot be empty");
+  }
+  if (failed(IREE::Codegen::WorkgroupMappingAttr::verifyAttrList(
+          forallOp.getContext(), forallOp.getLoc(), mappingAttr->getValue()))) {
+    return failure();
+  }
+  SmallVector<IREE::Codegen::WorkgroupMappingAttr> workgroupMappingAttrs =
+      llvm::map_to_vector(mappingAttr.value(), [](Attribute attr) {
+        return cast<IREE::Codegen::WorkgroupMappingAttr>(attr);
+      });
+  return workgroupMappingAttrs;
+}
+
+/// Get the permutation that represents the mapping of loop dimensions to
+/// process dimensions.
+SmallVector<int64_t>
+getMappingPermutation(ArrayRef<IREE::Codegen::WorkgroupMappingAttr> mapping) {
+  return invertPermutationVector(llvm::map_to_vector(mapping, [&](auto a) {
+    return static_cast<int64_t>(mapping.size() - 1) - a.getMappingId();
+  }));
+}
+
+/// Return the procId and nprocs to use for each of the distributed loops,
+/// derived from `hal.interface.workgroup.id/count`s.
+static FailureOr<
+    std::pair<SmallVector<OpFoldResult>, SmallVector<OpFoldResult>>>
```

Comment:
I think `SmallVector<std::pair<OpFoldResult, OpFoldResult>>` might be cleaner and more explicit that the length of each vector is the same. Also fewer `applyPermutation` calls.

Reply:
Edit: I see how this is used and keeping them as separate vectors does seem easier there, leaving this comment as optional.

Reply:
Thanks for the suggestions. I am going to leave it this way for now.